File: code/frontend/src/features/checklist/services/checklist-service.ts
import { apiClient } from '@/lib/api-client';
import type { ApiResponse } from '@/features/auth/types';
import type {
  BatchUpdateChecklistInput,
  ChecklistDto,
  ChecklistProgressDto,
  CreateChecklistInput,
  UpdateChecklistInput,
} from '../types';

export const checklistService = {
  getTaskChecklists: async (taskId: string): Promise<ChecklistDto[]> => {
    const res = await apiClient.get<ApiResponse<ChecklistDto[]>>(`/tasks/${taskId}/checklists`);
    return res.data.data;
  },

  getChecklistProgress: async (taskId: string): Promise<ChecklistProgressDto> => {
    const res = await apiClient.get<ApiResponse<ChecklistProgressDto>>(`/tasks/${taskId}/checklists/progress`);
    return res.data.data;
  },

  createChecklist: async (taskId: string, data: CreateChecklistInput): Promise<ChecklistDto> => {
    const res = await apiClient.post<ApiResponse<ChecklistDto>>(`/tasks/${taskId}/checklists`, data);
    return res.data.data;
  },

  updateChecklist: async (checklistId: string, data: UpdateChecklistInput): Promise<ChecklistDto> => {
    const res = await apiClient.put<ApiResponse<ChecklistDto>>(`/checklists/${checklistId}`, data);
    return res.data.data;
  },

  toggleChecklistComplete: async (checklistId: string, completed?: boolean): Promise<ChecklistDto> => {
    const res = await apiClient.patch<ApiResponse<ChecklistDto>>(`/checklists/${checklistId}/complete`, {
      completed,
    });
    return res.data.data;
  },

  reorderChecklist: async (checklistId: string, position: number): Promise<ChecklistDto> => {
    const res = await apiClient.patch<ApiResponse<ChecklistDto>>(`/checklists/${checklistId}/reorder`, {
      position,
    });
    return res.data.data;
  },

  deleteChecklist: async (checklistId: string): Promise<void> => {
    await apiClient.delete<ApiResponse<void>>(`/checklists/${checklistId}`);
  },

  batchUpdateChecklists: async (taskId: string, data: BatchUpdateChecklistInput): Promise<ChecklistDto[]> => {
    const res = await apiClient.put<ApiResponse<ChecklistDto[]>>(`/tasks/${taskId}/checklists/batch`, data);
    return res.data.data;
  },
};

File: code/backend/src/main/java/com/taskflow/modules/checklist/service/ChecklistService.java
package com.taskflow.modules.checklist.service;

import com.taskflow.modules.checklist.dto.BatchUpdateChecklistRequest;
import com.taskflow.modules.checklist.dto.ChecklistDto;
import com.taskflow.modules.checklist.dto.ChecklistProgressDto;
import com.taskflow.modules.checklist.dto.CreateChecklistRequest;
import com.taskflow.modules.checklist.dto.UpdateChecklistRequest;

import java.util.List;
import java.util.UUID;

/**
 * Domain Service interface for managing task checklists, subtask progress, reordering, and batch operations.
 */
public interface ChecklistService {

    /**
     * Creates a new checklist item for a task.
     *
     * @param userId  UUID identifier of the requesting user
     * @param taskId  UUID identifier of the target task
     * @param request creation payload
     * @return ChecklistDto representation of the created checklist item
     */
    ChecklistDto createChecklist(UUID userId, UUID taskId, CreateChecklistRequest request);

    /**
     * Retrieves all active checklist items for a task ordered by position.
     *
     * @param userId UUID identifier of the requesting user
     * @param taskId UUID identifier of the target task
     * @return list of ChecklistDto items
     */
    List<ChecklistDto> getTaskChecklists(UUID userId, UUID taskId);

    /**
     * Calculates the completion progress metrics for a task's checklist.
     *
     * @param userId UUID identifier of the requesting user
     * @param taskId UUID identifier of the target task
     * @return ChecklistProgressDto containing total, completed, and percentage metrics
     */
    ChecklistProgressDto getChecklistProgress(UUID userId, UUID taskId);

    /**
     * Updates an existing checklist item's title and completion status.
     *
     * @param userId      UUID identifier of the requesting user
     * @param checklistId UUID identifier of the target checklist item
     * @param request     update payload
     * @return updated ChecklistDto
     */
    ChecklistDto updateChecklist(UUID userId, UUID checklistId, UpdateChecklistRequest request);

    /**
     * Toggles or explicitly sets the completion status of a checklist item.
     *
     * @param userId      UUID identifier of the requesting user
     * @param checklistId UUID identifier of the target checklist item
     * @param completed   boolean status
     * @return updated ChecklistDto
     */
    ChecklistDto toggleChecklistComplete(UUID userId, UUID checklistId, Boolean completed);

    /**
     * Reorders a checklist item's position.
     *
     * @param userId      UUID identifier of the requesting user
     * @param checklistId UUID identifier of the target checklist item
     * @param position    new position index
     * @return updated ChecklistDto
     */
    ChecklistDto reorderChecklist(UUID userId, UUID checklistId, Double position);

    /**
     * Soft-deletes a checklist item.
     *
     * @param userId      UUID identifier of the requesting user
     * @param checklistId UUID identifier of the target checklist item
     */
    void deleteChecklist(UUID userId, UUID checklistId);

    /**
     * Performs a batch update or reorder across multiple checklist items for a task.
     *
     * @param userId  UUID identifier of the requesting user
     * @param taskId  UUID identifier of the target task
     * @param request batch update payload
     * @return list of updated ChecklistDto items
     */
    List<ChecklistDto> batchUpdateChecklists(UUID userId, UUID taskId, BatchUpdateChecklistRequest request);
}
